SLING-9053 Implement a retry mechanism for 503 or broken connection while AbstractSlingClient.doRequest(...) is called

tyge68 opened the pull request:
Define properties
-Dsling.it.http.retries : define the number of retries for 503 or broken connection , default to 5.
-Dsling.it.http.retriesDelay: define the number of milliseconds to wait before each retries. default to 1000 milliseconds.

Note depending on how parameters are set, a 503 might result in an exception or not, thus code need to either check the response code, or catch the exception. For broken connection it usually result in a RuntimeException which the message that it couldn't read content.

raducotescu commented:
I understand the need for retrying the requests, but I think that this should be handled at a lower level, with a retry handler [0], rather than directly in the `SlingClient` (cc @dulvac, @volteanu).

[0] - https://hc.apache.org/httpcomponents-client-4.5.x/httpclient/apidocs/org/apache/http/client/HttpRequestRetryHandler.html

tyge68 commented:
@raducotescu  I changed the implementation to use the http component api solution you mentioned, also I have implemented a logger for 503 which can be turned on with -Dsling.it.http.logRetries=true (default is using then the default implementation without logging detected 503. Logging the detected 503 is still nice as that can help teams to improve backend systems. Finally I added a test for it, if that help.

volteanu commented:
I agree it's more elegant with a retry strategy at HttpClient level, but I'm wondering if there are any cases for negative tests where we would actually expect a 503 (do wrong request and expect 503).

With a retry at `AbstractSlingClient#doRequest()` we could add some logic depending on the expected response status.

tyge68 commented:
@volteanu if the 503 is part of the expected test result, the implementation will still return the 503, just indeed it will retry a few time, but at least then it is a clear 503.

dulvac commented:
@tyge68 , thanks for the well-constructed contribution. I'm going to change it a little bit and apply it as a patch today.
